### ISSUE TITLE
Rebooted slave cannot send metrics to the master anymore

### ISSUE THREAD

**Issue by yasharne** [netdata/netdata#8930]
##### Bug report summary

One of our slaves restarted due to some problem and the netdata master gave up the last collected alerts so we realized there was something wrong with the slave (which was great) but the problem is after the reboot, the slave couldn't send metrics to the master, here is the error logs on the slave:
```2020-05-07 00:19:40: netdata INFO  : STREAM_SENDER[mariadb-07.db.x.y.z] : STREAM mariadb-07.db.x.y.z [send to 172.16.76.7:19999]: connecting...
2020-05-07 00:19:40: netdata INFO  : STREAM_SENDER[mariadb-07.db.x.y.z] : STREAM mariadb-07.db.x.y.z [send to 172.16.76.7:19999]: initializing communication...
2020-05-07 00:19:40: netdata INFO  : STREAM_SENDER[mariadb-07.db.x.y.z] : STREAM mariadb-07.db.x.y.z [send to 172.16.76.7:19999]: waiting response from remote netdata...
2020-05-07 00:19:40: netdata ERROR : STREAM_SENDER[mariadb-07.db.x.y.z] : STREAM mariadb-07.db.x.y.z [send to 172.16.76.7:19999]: server is not replying properly (is it a netdata?).
```
and here is the error logs on the master about this host:
```
2020-05-07 00:20:30: netdata INFO  : STREAM_RECEIVER[mariadb-07.db.x.y.z,[172.21.4.1]:13567] : thread created with task id 10106
2020-05-07 00:20:30: netdata INFO  : STREAM_RECEIVER[mariadb-07.db.x.y.z,[172.21.4.1]:13567] : set name of thread 10106 to STREAM_RECEIVER
2020-05-07 00:20:30: netdata INFO  : STREAM_RECEIVER[mariadb-07.db.x.y.z,[172.21.4.1]:13567] : STREAM mariadb-07.db.x.y.z [172.21.4.1]:13567: receive thread created (task id 10106)
2020-05-07 00:20:30: netdata INFO  : STREAM_RECEIVER[mariadb-07.db.x.y.z,[172.21.4.1]:13567] : STREAM mariadb-07.db.x.y.z [receive from [172.21.4.1]:13567]: multiple streaming connections for the same host detected. Rejecting new connection.
2020-05-07 00:20:30: netdata INFO  : STREAM_RECEIVER[mariadb-07.db.x.y.z,[172.21.4.1]:13567] : STREAM mariadb-07.db.x.y.z [receive from [172.21.4.1]:13567]: receive thread ended (task id 10106)
2020-05-07 00:20:30: netdata INFO  : STREAM_RECEIVER[mariadb-07.db.x.y.z,[172.21.4.1]:13567] : thread with task id 10106 finished
```

##### OS / Environment

master node:
```
Linux netdata-master-01.tool.x.y.z 4.15.0-76-generic #86-Ubuntu SMP Fri Jan 17 17:24:28 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
/etc/lsb-release:DISTRIB_ID=Ubuntu
/etc/lsb-release:DISTRIB_RELEASE=18.04
/etc/lsb-release:DISTRIB_CODENAME=bionic
/etc/lsb-release:DISTRIB_DESCRIPTION="Ubuntu 18.04.1 LTS"
/etc/os-release:NAME="Ubuntu"
/etc/os-release:VERSION="18.04.1 LTS (Bionic Beaver)"
/etc/os-release:ID=ubuntu
/etc/os-release:ID_LIKE=debian
/etc/os-release:PRETTY_NAME="Ubuntu 18.04.1 LTS"
/etc/os-release:VERSION_ID="18.04"
/etc/os-release:HOME_URL="https://www.ubuntu.com/"
/etc/os-release:SUPPORT_URL="https://help.ubuntu.com/"
/etc/os-release:BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
/etc/os-release:PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
/etc/os-release:VERSION_CODENAME=bionic
/etc/os-release:UBUNTU_CODENAME=bionic
```
Slave node:
```
Linux mariadb-07.db.x.y.z 4.4.0-151-generic #178-Ubuntu SMP Tue Jun 11 08:30:22 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
/etc/lsb-release:DISTRIB_ID=Ubuntu
/etc/lsb-release:DISTRIB_RELEASE=16.04
/etc/lsb-release:DISTRIB_CODENAME=xenial
/etc/lsb-release:DISTRIB_DESCRIPTION="Ubuntu 16.04.6 LTS"
/etc/os-release:NAME="Ubuntu"
/etc/os-release:VERSION="16.04.6 LTS (Xenial Xerus)"
/etc/os-release:ID=ubuntu
/etc/os-release:ID_LIKE=debian
/etc/os-release:PRETTY_NAME="Ubuntu 16.04.6 LTS"
/etc/os-release:VERSION_ID="16.04"
/etc/os-release:HOME_URL="http://www.ubuntu.com/"
/etc/os-release:SUPPORT_URL="http://help.ubuntu.com/"
/etc/os-release:BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
/etc/os-release:VERSION_CODENAME=xenial
/etc/os-release:UBUNTU_CODENAME=xenial
```
##### Netdata version

master: `netdata v1.21.1`
slave: `netdata v1.19.0`
 

##### Component Name

stream

**Comment by thiagoftsm** [comment]
Hi @yasharne ,

During the slave reboot, could you see in the logs if the slave was closed graceful?  I am thinking that Netdata could not close correctly the stream and the master is not allowing it to reconnect.

Another doubt, how long the slave wait for to reconnect  with Netdata? It is possible we need to adjust this.

Best regards!

**Comment by yasharne** [comment]
Hi @thiagoftsm 
the error logs were not enabled before the reboot so I cannot see the logs but according to our NOC report, it wasn't graceful (the os was crashed)
The reconnect delay is 5 seconds

**Comment by thiagoftsm** [comment]
This explains the problem!

@mfundul, @amoss, @vlvkobal we will need to bring changes in our code to accept new connections when events like this happens.

**Comment by csch42** [comment]
I have been observing similar behavior. We are using netdata on routers and some of them (mobile ones, using LTE) are observing "unexpected" link-downs. After re-connect the slave on the router cannot re-establish a stream to the master for ~ 2 hours. Wiresharking the connection reveals, that the slave _can_ establish a TCP connection, but immediately after sending his initial "greeting", the server drops the connection again. I guess it says "Hi, I am XYZ ..." and the server goes "Naaah ... I already have an existing connection to XYZ, go away, imposter!" and drops the TCP connection?

**Comment by ilyam8** [comment]
@csch42 m/s version? latest?

**Comment by cakrit** [comment]
I see the slave is v1.19, but that doesn't matter. The master should have realized that it's not receiving metrics from the slave with the particular GUID any more and immediately accept the new connection.

**Comment by csch42** [comment]
Master is v1.19.0-258-nightly, slaves are of various versions, for example also netdata v1.19.0-258-nightly

**Comment by ilyam8** [comment]
i think it could help, if you have it please share

**Comment by csch42** [comment]
I'll create one later today (force a network disruption).

**Comment by mfundul** [comment]
The master cannot accept metrics from the slave until it makes sure the existing thread serving the previous connection is invalid. This is the logic that can fix this issue.

**Comment by csch42** [comment]
Sorry it took so long. "Real life" intervened with other issues ... 

Also: I am not publicly posting the pcap (for obvious reasons), but I'll happily provide the raw pcap via DM, if needed.

However, it doesn't show much. You see the otherwise uneventful TCP-SYN/ACK exchange (my setup uses IPv6, btw.), then the streaming client sends its hello:

```
STREAM key=ab7c5a67-bd32-4062-872c-186703d0a41e&hostname=M79-RT-A&registry_hostname=M79-RT-A&machine_guid=c0b22798-49a6-11ea-b0c5-18e8294edc60&update_every=1&os=linux&timezone=Europe/Berlin&tags=&NETDATA_SYSTEM_OS_NAME=Debian GNU/Linux&NETDATA_SYSTEM_OS_ID=debian&NETDATA_SYSTEM_OS_ID_LIKE=unknown&NETDATA_SYSTEM_OS_VERSION=9 (stretch)&NETDATA_SYSTEM_OS_VERSION_ID=9&NETDATA_SYSTEM_OS_DETECTION=/etc/os-release&NETDATA_SYSTEM_KERNEL_NAME=Linux&NETDATA_SYSTEM_KERNEL_VERSION=4.9.79-UBNT&NETDATA_SYSTEM_ARCHITECTURE=mips64&NETDATA_SYSTEM_VIRTUALIZATION=unknown&NETDATA_SYSTEM_VIRT_DETECTION=none&NETDATA_SYSTEM_CONTAINER=unknown&NETDATA_SYSTEM_CONTAINER_DETECTION=none HTTP/1.1
User-Agent: netdata/v1.19.0-258-nightly
Accept: */*

```

... at which point the server, without any hesitation or comment, immediately closes the TCP connection with the usual TCP-FIN/ACK farewell.

Five seconds later: Wash, rinse, repeat. But after two hours and a few seconds the two endpoints start talking to each other, again.

**Comment by amoss** [comment]
Looks like the same problem as #8825 - fix will be the same PR, estimated to be middle of next week.

**Comment by amoss** [comment]
We've found the problem - it was the stability fixes that were added in March to handle large numbers of slaves. They had an unexpected interaction with the code that rejects multiple connections from the sender. The fix is being delayed slightly because it interacts with another piece of work that is in progress (fixing the gaps in charts when slaves restart). The PR should be complete by the end of next week - I'll let you know when it is finished.
